[ENH] Restructuring of DWI derivatives (#5)
bids-standard/bids-bep016: src/05-derivatives/05-diffusion-derivatives.md

```

The JSON sidecar file is REQUIRED (due to the REQUIRED `SkullStripped` field -
see [Common Data Types](02-common-data-types.md)), and MAY additionally be used to
store information about what preprocessing options were used (for example
whether denoising was performed, corrections applied for field inhomogeneity /
gradient non-linearity / subject motion / eddy currents, etc.).

Additional reserved JSON metadata fields:

> | **Key name**                           | **Description**                                                                                                                                   |

Reviewer: +1

> | **Key name**                           | **Description**                                                                                                                                   |

Author: Given https://github.com/bids-standard/bids-specification/pull/300, are we in fact better off omitting provenance information entirely?

> | **Key name**                           | **Description**                                                                                                                                   |

Reviewer 2: I think we should omit provenance information, else the spec will get even more complicated, and will always lag behind when new developments / updates to algorithms are made.

> | **Key name**                           | **Description**                                                                                                                                   |

Reviewer 2: Or, at least, I think most of those fields should simply be `Boolean`. That way, the user can quickly know if, for example, motion correction was done, without imposing a vocabulary to method / pipelines developers.

From my POV, I would keep the "general" options, such as `Denoising`, `GibbsRingingCorrection`, but would simplify the fields like `GradientNonLinearity*Correction` to only `GradientNonLinearityCorrection`, as a boolean. If the end user needs more details on the exact methods that were used, a provenance system is probably more suited to those details.

| BiasFieldCorrectionMethod              | OPTIONAL. String. Method (if any) used for correction of B1 RF field inhomogeneity                                                                |

## Diffusion models

Diffusion MRI can be modeled using various paradigms to extract more
informative representations of the diffusion process and the underlying
biological structure. A wide range of such models are available, each of
which has its own unique requirements with respect to:

-   The input parameters required to define / constrain the model;

-   The appropriate data representation in which the resulting
    parameters estimated of / from the model are stored on the filesystem;

-   The requirements for encapsulation and complete representation of
    derived orientation information, which is a key strength of diffusion
    MRI but presents unique challenges for correct interpretation.

### Parameter terminology

1.  *Intrinsic* parameter: Item that is the direct result of fitting
    the diffusion model to the empirical diffusion-weighted data.

1.  *Input* parameter: Item that influences the conformation of the
    diffusion model to the empirical diffusion-weighted data.

> 1.  *Extrinsic* parameter: Item that can be calculated from prior

Author: Every time I tried to more stringently separate intrinsic and extrinsic parameters it always felt artificial. Even though they're *derived* differently, their actual representation on disk, and potential subsequent utility, isn't actually really that much affected by whether or not "the DWI data were being referred to in the generation of the parameter". So there would end up being two different definitions and file naming conventions for two pieces of data that were otherwise "equals". It can even get tricky whereby a particular software tool can export a parameter as part of the fitting procedure that's actually extrinsic, but was generated and saved along with intrinsic parameters, either out of convenience or for a lack of dedicated stand-alone command for generating such or even for simply not being interested in the distinction.

I don't see that fitting a model with one tool and extracting extrinsic parameters with another is incompatible with the current proposal. There's one JSON across all intrinsic model parameters that can contain the software used to fit the model, but each extrinsic parameter can additionally have its own JSON, which could contain the software used to extract that particular parameter if desired.

I feel like the "modality" as you put it, ie. the final key in the file name (is that a common nomenclature for it?), should have the minimal cardinality possible across BIDS. Adding one *per parameter* across all possible diffusion model parameters would very much affect that.

> 1.  *Extrinsic* parameter: Item that can be calculated from prior

Reviewer 3: Following-up here #12.

    estimated intrinsic model parameters, without necessitating reference
    to the empirical diffusion-weighted data.

> ### File names

Author: I struggled with the order of sections as well. I think it's difficult to move "intrinsic model parameters" above the "data representations" section, since the former contains so many links to the latter. The only thing in between them then is "orientation specification": while it would be 'safe' to have this below "intrinsic model parameters", it itself is reasonably tightly linked to "data representations". This would also move "intrinsic model parameters" further away from "input parameters" and "extrinsic model parameters"; which are we've seen from this process are quite strongly related also. I understand the motivation to get the model definitions up as high as possible, I experienced the same thing.

I think if this were to change, it would not be "moving the table of parameters up" so much as "moving data representations & orientation specification to the end (just before the examples)".

So instead of:

Diffusion derivatives
- Preprocessed diffusion-weighted images
- Diffusion models
  - Parameter terminology
  - File names
  - Data representations
  - Orientation specification
    - Spherical Harmonics bases
  - Intrinsic model parameters
    - Model bootstrapping
  - Input model parameters
  - Extrinsic model parameters
- Demonstrative examples

, it would instead be:

Diffusion derivatives
- Preprocessed diffusion-weighted images
- Diffusion models
  - Parameter terminology
  - File names
  - Intrinsic model parameters
    - Model bootstrapping
  - Input model parameters
  - Extrinsic model parameters
  - Data representations
  - Orientation specification
    - Spherical Harmonics bases
- Demonstrative examples

It would mean that "data representations" would get referred to *a lot* before actually appearing though.

> ### File names

Author: Looking at the proposed change itself: I do think it would be useful to have a brief introduction paragraph at the head of some of the sections. But in looking to adopt / modify your suggestion here, I'm still not happy with the overall "file names" section; I also think some of what you were starting to get at in that text was not wholly specific to "file names". So what I might do instead is eventually decide on a different structure for what's currently the "file names" section, as the requisite modifications from that point may be entirely different.

> ### File names

Author: See 8b27828 where "orientation specification" is moved down, concomitantly moving the parameter tables up.

```Text
<pipeline_name>/
    sub-<participant_label>/
        dwi/

>             <source_keywords>[_space-<space>][_desc-<label>]_parameter-<intparam1>_<model>.nii[.gz]

Author: I flip-flopped a bit while writing this with not specifically your point here, but more generally with whether the document should read like a tutorial or a specification; my first draft was more like the former, and I deleted a *lot* of text to converge toward the latter. I'll add a link to the later section where the permissible model names are listed; but "explaining" or "giving examples" at this point in the page would actually I think opens Pandora's Box.

>             <source_keywords>[_space-<space>][_desc-<label>]_parameter-<intparam1>_<model>.nii[.gz]

Author: See if 6cf8a62 is adequate.

            <source_keywords>[_space-<space>][_desc-<label>]_parameter-<intparam2>_<model>.nii[.gz]
            <source_keywords>[_space-<space>][_desc-<label>]_<model>.json
            [<source_keywords>[_space-<space>][_desc-<label>]_parameter-<extparam1>_<model>.nii[.gz]]
            [<source_keywords>[_space-<space>][_desc-<label>]_parameter-<extparam1>_<model>.json]
            [<source_keywords>[_space-<space>][_desc-<label>]_parameter-<extparam2>_<model>.nii[.gz]]
            [<source_keywords>[_space-<space>][_desc-<label>]_parameter-<extparam2>_<model>.json]
```

-   File "`<source_keywords>[_space-<space>][_desc-<label>]_<model>.json`"
    provides basic model information and [input model parameters](#input-model-parameters).

-   OPTIONAL file "`<source_keywords>[_space-<space>][_desc-<label>]_parameter-<extparam*>_<model>.json`"
    provides only information or parameters relevant to derivation of the
    relevant [extrinsic model parameter](#extrinsic-model-parameters) based

>     on prior estimates of [intrinsic model parameters](#intrinsic-model-parameters).

Author: Fundamentally the question becomes:

1. How far do two different pieces of data need to differ in their position in the hierarchy of image processing / derivation before they need to be presented explicitly as existing at different positions within that hierarchy?

2. In what way should this variation in position within the hierarchy be encoded?

Clearly we have a distinction between BIDS raw data and BIDS derivatives, which is defined by having a sub-folder for the processing pipeline that produced the derivative. Does the "progression" from intrinsic model parameters to extrinsic model parameters also warrant a bifurcation in order to codify that dependence; and if so, how "large" should the distinction between them be?

I struggled a lot with this when coming up with the proposal, and kinda wish I'd written down my thoughts at the time. But I do think that the fact that some models can blur the lines slightly between intrinsic and extrinsic means that the distinction between the two should be on the smaller side.

As an example: Consider the fibre orientations output by `bedpostx`. The most "native" form of its output (talking specifically about the mean across bootstraps here) is the two spherical angles and fibre volume fraction per stick per voxel. It also outputs the "dyads" images, which are 3-vectors that also encode the fibre orientations. Now the latter are exported by `bedpostx`, so could be considered intrinsic parameters. But if these data were *not* exported by `bedpostx`, it would not be difficult to generate these data from the spherical angle images. Is this therefore then extrinsic parameter data?

An alternative way of looking at it: If examining the contents of a BIDS derivatives directory, it may not actually be that consequential to be able to easily identify which images are intrinsic and which are extrinsic. If the distinction is important in a particular case, then it's likely that JSON files would need to be provided for both, at which point the distinction can be inferred. But if just looking at the data, which are intrinsic and which are extrinsic don't actually matter that much. If looking to detect & automate the further processing of such, that will require knowledge of the spec, at which point those parameters considered intrinsic are known from the spec.

So: While technically yes there is a "derivative" dependence here, I'm not sure that codifying such would actually be beneficial pragmatically, and may in fact come with detriments. Happy for possible alternative solutions to be floated though.

#### Spherical Harmonics bases

-   `MRtrix3`

    -   Antipodally symmetric: all basis functions with odd degree are
        assumed zero; `AntipodalSymmetry` MUST NOT be set to True.

    -   Functions assumed to be real: conjugate symmetry is assumed, i.e.
        *Y*(*l*,-*m*) = *Y*(*l*,*m*)\*, where \* denotes the complex
        conjugate.

    -   Mapping of image volumes to spherical harmonic basis function
        coefficients:

        | **Volume** | **Coefficient**                   |
        | ---------- | --------------------------------- |
        | 0          | *l* = 0, *m* = 0                  |
        | 1          | *l* = 2, *m* = 2 (imaginary part) |
        | 2          | *l* = 2, *m* = 1 (imaginary part) |
        | 3          | *l* = 2, *m* = 0                  |
        | 4          | *l* = 2, *m* = 1 (real part)      |
        | 5          | *l* = 2, *m* = 2 (real part)      |
        | 6          | *l* = 4, *m* = 4 (imaginary part) |
        | 7          | *l* = 4, *m* = 3 (imaginary part) |
        | ...        | etc.                              |

>     -   Normalisation: ***TODO***

Author: Decided to leave out the MRtrix 0.2 convention; that software is likely almost never used, and the orthonormal basis in *MRtrix3* has been the default for many years now. Indeed we're considering removing support for the old non-orthonormal basis entirely in *MRtrix3*.

What this statement more refers to is that the spec requires a description regarding how the overall magnitude of the SH coefficients should be interpreted. E.g. If the *l*=0 term is 1.0 and all other coefficients are 0.0, does this mean that the FOD amplitude is 1.0 in all directions? Or that the integral over the half-sphere / full-sphere is 1.0? E.g. see: [Wikipedia](https://en.wikipedia.org/wiki/Spherical_harmonics#Orthogonality_and_normalization).

I got Donald to revise his documentation around this standard in https://github.com/MRtrix3/mrtrix3/pull/1635; so I will look to update this spec accordingly.

>     -   Normalisation: ***TODO***

Reviewer 3: @Lestropie did you finally update the PR accordingly?

>     -   Normalisation: ***TODO***

Author: Moved this to #8, which currently has the branch of this PR as the target.

>     -   Normalisation: ***TODO***

Author: Having said that, I put that in a separate PR in order to provide a clean online diff, but it seems to be off somehow. I'll have to figure out what's happened with the force pushes to `dwi_derivatives_restructure`.

    Relationship between maximal spherical harmonic degree *l<sub>max</sub>*
    and number of image volumes *N*:

        *N* = ((*l<sub>max</sub>*+1) x (*l<sub>max</sub>*+2)) / 2

        | ***l<sub>max</sub>*** | ***N*** |
        | --------------------- | ------- |
        | 0                     | 1       |
        | 2                     | 6       |
        | 4                     | 15      |
        | 6                     | 28      |
        | 8                     | 45      |
        | ...                   | etc.    |

    Relationship between maximal degree of *zonal* spherical harmonic
    function (spherical harmonics function where all *m* != 0 terms are
    assumed to be zero; used for e.g. response function definition) and
    number of coefficients *N*:

        *N* = 1 + (*l<sub>max</sub>* / 2)

        | ***l<sub>max</sub>*** | ***N*** |
        | --------------------- | ------- |
        | 0                     | 1       |
        | 2                     | 2       |
        | 4                     | 3       |
        | 6                     | 4       |
        | 8                     | 5       |
        | ...                   | etc.    |

>  -   `Descoteaux`

Author: As discussed: likely what will be most succinct is a description of how this convention differs from the *MRtrix3* one. Many of the details will be shared between the two.

>  -   `Descoteaux`

Reviewer 2: I have the content, I just need to format it correctly. Do you want to get a PR on your PR? Or I just past it in a comment and you update it yourself?

>  -   `Descoteaux`

Author: While it's possible to make proposed changes to a PR within a comment, I think I'd rather strip it from this PR, merge, and then we can start making individual PRs for individual changes like this. It will also make the relevant diffs far easier to navigate. We shouldn't see this individual PR as the "final" BEP.

The JSON sidecar for the intrinsic diffusion model parameters may contain
the following key/value pairs irrespective of the particular model:

| **Key name**     | **Description**                                                                                                                                                                              |
| ---------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | Gradients        | OPTIONAL. List of 3-vectors. Subset of gradients utilized to fit the model, as a list of three-elements lists. If not present, all gradients were used.                                      |

Reviewer 3: ```suggestion
| Gradients        | REQUIRED. List of 3-vectors. Subset of gradients utilized to fit the model, as a list of three-elements lists.                                                                               |
```

Reviewer 3: If these two fields are not required we will easily end up having all models use all the gradients always.

Author: I would expect the majority of model implementations to be compatible with using all data, as well as all experimental uses of such to have utilised all data. So the necessity to clarify which of the input data were used should in fact be quite rare. What could be said instead is that the field is REQUIRED if the complete set of input DWI data were not used.

See also concern RE: `Gradients` and `Shells` fields in first PR comment.

> | Shells           | OPTIONAL. List of floats. Shells that were utilized to fit the model, as a list of b-values. If the key is not present, it should be assumed that all shells were used during model fitting. |

Reviewer 3: ```suggestion
| Shells           | REQUIRED. List of floats. Shells that were utilized to fit the model, as a list of b-values.                                                                                                 |
```

Reviewer 3: If the above is `REQUIRED`, is this field even necessary?

Reviewer 2: I think, as @Lestropie mentionned initially, we need to really specify the utilized combination, since people could use some gradients from some shells, especially if the same direction is present on multiple shells.

## Diffusion models

Diffusion MRI can be modeled using various paradigms to extract more
informative representations of the diffusion process and the underlying
biological structure. A wide range of such models are available, each of
which has its own unique requirements with respect to:

-  The [input](#paramdef-input) parameters required in order to
   define / constrain the model;

-  The appropriate [data representations](#data-representations) utilised to
   store information parameterised [by](#paramdef-intrinsic) or
   [from](#paramdef-extrinsic) the model onto the filesystem;

-  The requirements for encapsulation and complete representation of
   derived [orientation information](#orientation-specification), which is
   a key strength of diffusion MRI but presents unique challenges for
   correct interpretation.

### Parameter terminology

Throughout this document, the term "parameter" is used to refer to
multiple distinct sources of information. The distinction between
these uses is defined thus:

1. <a name="paramdef-input">*Input*</a> model parameter:

   Value or non-numerical setting that influences the conformation

>    of the diffusion model to the empirical diffusion-weighted data.

Reviewer 2: Could we add a simple example of each of the 3 types of parameters so that first time readers of the spec who have a dMRI knowledge can quickly map them in their mind? For example, saying that the SH order is an input parameter.

Author: I think a lot of the pre-processed data fields may end up disappearing entirely due to not wanting to manage provenance within BIDS.

> Could we add a simple example of each of the 3 types of parameters

Yes, I think throwing in an example here could be beneficial.

Author: > > Could we add a simple example of each of the 3 types of parameters

> Yes, I think throwing in an example here could be beneficial.

Added in 1c7b67f.

   intrinsic model parameters, without necessitating reference to
   the empirical diffusion-weighted data.

### File names

```Text
<pipeline_name>/
    sub-<participant_label>/
        dwi/
            <source_keywords>[_space-<space>][_desc-<label>]_parameter-<intparam1>_<model>.nii[.gz]
            <source_keywords>[_space-<space>][_desc-<label>]_parameter-<intparam2>_<model>.nii[.gz]
            <source_keywords>[_space-<space>][_desc-<label>]_<model>.json

            [<source_keywords>[_space-<space>][_desc-<label>]_parameter-<extparam1>_<model>.nii[.gz]]
            [<source_keywords>[_space-<space>][_desc-<label>]_parameter-<extparam1>_<model>.json]
            [<source_keywords>[_space-<space>][_desc-<label>]_parameter-<extparam2>_<model>.nii[.gz]]
            [<source_keywords>[_space-<space>][_desc-<label>]_parameter-<extparam2>_<model>.json]
```

-  File "`<source_keywords>[_space-<space>][_desc-<label>]_<model>.json`"
   provides basic model information and [input](#paramdef-input)
   model parameters.

-  Images "`<source_keywords>[_space-<space>][_desc-<label>]_parameter-<intparam*>_<model>.nii[.gz]`"

>    MAY be defined, in order to provide additional [extrinsic](#paramdef-extrinsic)

Reviewer 2: Should `extrinsic` be `intrinsic` here? Or, if not the case, then the filename should refer to `extparam`. Either way I think we are missing either the `extparam` or `intparam` example.

Author: Corrected in 90fa5fa.

>    MAY be defined, in order to provide additional [extrinsic](#paramdef-extrinsic)

Author: The mistake is in the file name, not the subsequent text. It's intended to refer to extrinsic parameters.

>    MAY be defined, in order to provide additional [extrinsic](#paramdef-extrinsic)

Author: Resolved by 90fa5fa.

   model parameters.

-  OPTIONAL files "`<source_keywords>[_space-<space>][_desc-<label>]_parameter-<extparam*>_<model>.json`"
   may provide only information or parameters relevant to derivation of each
   relevant [extrinsic](#paramdef-extrinsic) model parameter image.

> -  If all [intrinsic](#paramdef-intrinsic) model parameters are

Author: Possibly. I've never been entirely happy with this separation of file naming convention between one and multiple NIfTI images for intrinsic model parameters. Open to opinions one way or the other.

> -  If all [intrinsic](#paramdef-intrinsic) model parameters are

Author: Decided I preferred this alternative; change in 76cf83d.

> -  If all [intrinsic](#paramdef-intrinsic) model parameters are

Reviewer 2: Sounds good to me.

   contained within a single image file, field "`_parameter-*`" MUST be
   omitted (for intrinsic parameter image only):

   ```Text
   <pipeline_name>/
       sub-<participant_label>/
           dwi/
               <source_keywords>[_space-<space>][_desc-<label>]_<model>.nii[.gz]
               <source_keywords>[_space-<space>][_desc-<label>]_<model>.json
   ```

### Data representations

There are multiple techniques by which data that relate to the anisotropic
nature of either the diffusion process or underlying tissue may be
arranged and/or encoded into NIfTI image data. A list of known techniques
is enumerated below, accompanied by requisite information specific to the
reading / writing of each representation.

1. <a name="data-scalar">*Scalars*</a>:

   Any model parameter image (whether [intrinsic](#paramdef-intrinsic) or
   [extrinsic](#paramdef-extrinsic)) where a solitary numerical value is
   defined in each 3D image voxel is referred to here as a "scalar" image.

1. <a name="data-dec">*Directionally-Encoded Colours (DEC)*</a>:

>    4D image with three volumes, intended to be interpreted as red, green

Reviewer 2: Here, I know that we had some discussions on this during the initial derivatives meeting. This format is what most tools now use, but is not the official NIfTI way (using a special datatype and intent code). I think we should probably support both, since, for example, Trackvis still expects the official NIfTI standard.

Author: This was raised during the meeting; I was not aware of this format previously. Part of the issue is that many softwares ignore NIfTI intent fields. It would theoretically be possible to support both; but it would I think at least require identification of use of this format within the JSON sidecar rather than relying on the NIfTI intent code.

Reviewer 2: I agree that most software just disregard intent codes. I think identifying it in the sidecar is a good idea, but I would still allow using the intent code for tools that support it natively.

         with value of 0.0 corresponding to the *first* reference axis
         (see [orientation specification](#orientation-specification)),
         increasing according to the right-hand rule about the zenith
         direction.

      Number of image volumes is equal to (3x*N*), where *N* is the maximum
      number of discrete orientations in any voxel in the image.

   1. Directions only

      Each consecutive pair of image volumes encodes inclination /
      azimuth pairs, with order & convention identical to that above
      (equivalent to spherical coordinate with assumed unity distance from
      origin).

      Number of image volumes is equal to (2x*N*), where *N* is the maximum
      number of discrete orientations in any voxel in the image.

1. <a name="data-3vector">*3-Vectors*</a>:

   4D image where data across volumes within each voxel encode one or
   more discrete orientations using triplets of axis dot products.

   This representation may be used in one of two ways:

   1. Value per direction

      Each 3-vector, once explicitly normalized, provides a direction
      on the unit sphere; the *norm* of each 3-vector additionally encodes

>       the magnitude of some mode parameter, the nature of which MUST be

Reviewer 2: Did you intend on writing `model` instead of `mode`?

Author: Fixed in 6608713.

| --------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `AntipodalSymmetry`         | [Spherical coordinates](#data-spherical), [3-vectors](#data-3vector), [spherical harmonics](#data-sh), [amplitudes](#data-amp), [probability distribution functions](#data-pdf), [parameter vectors](#data-param) | OPTIONAL. Boolean. Indicates whether orientation information should be interpreted as being antipodally symmetric. Assumed to be True if omitted.                                                                                                         |

> | `Directions`                | [Amplitudes](#data-amp)                                                                                                                                                                                           | REQUIRED. List. Data are either [spherical coordinates (directions only)](#data-spherical) or [3-vectors](#data-3vector) with unit norm. Defines the dense directional basis set on which samples of a spherical function within each voxel are provided. |

Author: Possibly. In part I went with `Directions` because *MRtrix3* already supports this, and at least some commands are already compatible with either representation. Then again, the convention for spherical angles in *MRtrix3* is incompatible with what I've put into the spec, which is instead based on ISO. So maybe using different keys would actually be preferable.

> | `Directions`                | [Amplitudes](#data-amp)                                                                                                                                                                                           | REQUIRED. List. Data are either [spherical coordinates (directions only)](#data-spherical) or [3-vectors](#data-3vector) with unit norm. Defines the dense directional basis set on which samples of a spherical function within each voxel are provided. |

Reviewer 3: What is the status of this question?

> | `Directions`                | [Amplitudes](#data-amp)                                                                                                                                                                                           | REQUIRED. List. Data are either [spherical coordinates (directions only)](#data-spherical) or [3-vectors](#data-3vector) with unit norm. Defines the dense directional basis set on which samples of a spherical function within each voxel are provided. |

Author: I think I'd like to switch to having two separate keys. Though personally my instinct was `DirectionsCartesian` and `DirectionsSpherical`. Does this clash too much with the "`3-vectors`" data representation?

> | `Directions`                | [Amplitudes](#data-amp)                                                                                                                                                                                           | REQUIRED. List. Data are either [spherical coordinates (directions only)](#data-spherical) or [3-vectors](#data-3vector) with unit norm. Defines the dense directional basis set on which samples of a spherical function within each voxel are provided. |

Reviewer 3: Follow up in #18.

| `SphericalHarmonicDegree`   | [Spherical harmonics](#data-sh)                                                                                                                                                                                   | REQUIRED. Integer. Maximal degree of the spherical harmonic basis employed.                                                                                                                                                                               |

If `AntipodalSymmetry` is True, then no constraints are imposed with respect
to the domain on the 2-sphere in which orientations may be specified;
for instance, 3-vectors { 0.57735, 0.57735, 0.57735 } and
{ -0.57735, -0.57735, -0.57735 } are both permissible and equivalent to one
another.

> #### Spherical Harmonics bases

Reviewer 2: I feel like this should probably be a kind of appendix, or put at the end of the document. Most end users will only need to know which basis was used, to tell their tools (if needed) which basis to use.

Author: Related to [this comment](https://github.com/bids-standard/bids-bep016/pull/5#discussion_r314176260). In the case of the reordering I proposed there, the whole "orientation specification" section would end up being the last section prior to the demonstrative examples, which would indirectly give the same result as suggested here.

Author: See 8b27828.

| ----------- | ----------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `bs`        | Ball-and-Stick(s) model \[[Behrens2003](#behrens2003)\],\[[Behrens2007](#behrens2007)\],\[[Jbabdi2012](#jbabdi2012)\]                           | One [spherical coordinates](#data-spherical) image with parameter name "`sticks`", providing both fibre volume fractions and orientations using polar angles;<br>Optional scalar images with parameter names {"`bzero`", "`dmean`", "`dstd`"} providing the model-estimated *b*=0 signal intensity, mean stick diffusivity, and standard deviation of stick diffusivities respectively                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |

Reviewer 4: Good choice of acronym!

Author: :laughing: 

I've still not gotten this to the point where I'm happy with it. Really, what we want to be encoding here is not specifically "ball" and "sticks", but the most general case of discrete mixture models that use different mathematical models for the different components (e.g. zeppelin(s), rackets), of which FSL bedpostx output / "ball-and-sticks" are merely specific instances of such. But I've not actually used such models, I'm only approaching this from a theoretical perspective. Would be interested in interacting with FMRIB or anyone who has experimented with the breadth of this class of discrete mixture models.

| `csd`       | Constrained Spherical Deconvolution \[[Tournier2007](#tournier2007)\],\[[Descoteaux2009](#descoteaux2009)\],\[[Jeurissen2014](#jeurissen2014)\] | [Spherical harmonics](#data-sh) image<br>If a multi-tissue decomposition is performed, provide one individual 4D image per tissue, with "`_desc-<desc>`" filename field being an abbreviation of the tissue estimated by that particular ODF                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |

> | `dki`       | Diffusion Kurtosis Imaging \[[Jensen2005](#jensen2005)\]                                                                                        | Single [parameter vectors] image with 21 volumes in the order: *D<sub>xx</sub>*, *D<sub>xy</sub>*, *D<sub>xz</sub>*, *D<sub>yy</sub>*, *D<sub>yz</sub>*, *D<sub>zz</sub>*, *W<sub>xxxx</sub>*, *W<sub>yyyy</sub>*, *W<sub>zzzz</sub>*, *W<sub>xxxy</sub>*, *W<sub>xxxz</sub>*, *W<sub>xyyy</sub>*, *W<sub>yyyz</sub>*, *W<sub>xzzz</sub>*, *W<sub>yzzz</sub>*, *W<sub>xxyy</sub>*, *W<sub>xxzz</sub>*, *W<sub>yyzz</sub>*, *W<sub>xxyz</sub>*, *W<sub>xyyz</sub>*, *W<sub>xyzz</sub>* (*D* is the diffusion tensor, *W* is the kurtosis tensor)<br>OR<br>6 diffusion tensor coefficients as [parameter vectors](#data-param) image with parameter name "`tensor`";<br>15 kurtosis tensor coefficients as [parameter vectors](#data-param) image with parameter name "`kurtosis`";<br>Optional: estimated *b*=0 intensity as scalar image with parameter name "`bzero`"                                                                              |

Author: Resoved by f64bfe8.

> | `dti`       | Diffusion Tensor Imaging \[[Basser1994](#basser1994)\]                                                                                          | Single [parameter vectors](#data-param) image with 6 volumes in the order: *D<sub>xx</sub>*, *D<sub>xy</sub>*, *D<sub>xz</sub>*, *D<sub>yy</sub>*, *D<sub>yz</sub>*, *D<sub>zz</sub>*<br>OR<br>Tensor coefficients as [parameter vectors] image with parameter name "`tensor`";<br>Estimated *b*=0 intensity as [scalar](#data-scalar) image with parameter name "`bzero`"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |

Reviewer 4: This one needs some rethinking. The NIFTI standard has a *different* way of encoding this kind of data, see: https://github.com/nipy/dipy/issues/1890. We might want to use that one, since it's the standards-compliant way. I don't think this generalizes to DKI, though.

Author: As discussed: we may not necessarily want to support the NIfTI standard format for encoding such data, since the NIfTI specification contains many other things that we likely more explicitly don't want to support. But it may depend on whether or not any softwares actually provide tensor data using this specification; if such are rare or non-existent, then I'd personally prefer to write this off under the 80/20 rule.

Reviewer 3: Has this been addressed? If not, please open a new issue for it.

Author: My personal preference is to define tensor data in a way we devise that is consistent with the rest of BIDS Derivatives, rather than trying to propagate aspects of the NIfTI format that are not themselves consistently utilised. Conversion between the two should not be difficult in terms of volume permutation, but I don't know what proportion of softwares would actually permit setting / obeying NIfTI intent codes. I'm not against including support for such, but it would need to be exceptionally clear in the JSON that that specific format is utilised, rather than relying on the NIfTI intent codes alone.

This one's slightly complex so happy to defer to a dedicated issue.

Reviewer 3: Dedicated issue is #17

> | `dti`       | Diffusion Tensor Imaging \[[Basser1994](#basser1994)\]                                                                                          | Single [parameter vectors](#data-param) image with 6 volumes in the order: *D<sub>xx</sub>*, *D<sub>xy</sub>*, *D<sub>xz</sub>*, *D<sub>yy</sub>*, *D<sub>yz</sub>*, *D<sub>zz</sub>*<br>OR<br>Tensor coefficients as [parameter vectors] image with parameter name "`tensor`";<br>Estimated *b*=0 intensity as [scalar](#data-scalar) image with parameter name "`bzero`"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |

Author: Resolved by f64bfe8.

| `forecast`  | Fiber ORientation Estimated using Continuous Axially Symmetric Tensors \[[Zuchelli2017](#zuchelli2017)\]                                        | [Spherical harmonics](#data-sh) image                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |

> | `fwdti`     | Free water DTI \[[Hoy2015](#hoy2015)\]                                                                                                          | One [p

Reviewer 3: Has this comment been addressed?

> | `fwdti`     | Free water DTI \[[Hoy2015](#hoy2015)\]                                                                                                          | One [p

Author: This text came from the original document prior to my hatchet job; so I don't actually know the answer without doing the relevant investigation myself.

> | `fwdti`     | Free water DTI \[[Hoy2015](#hoy2015)\]                                                                                                          | One [p

Reviewer 3: Thanks, added a new issue for this.

| `rtop`              | Return To Origin Probability                                           | [Scalar](#scalar)       | { `shore` }                                     | Probability \[0.0-1.0\]                                        |
| `rtpp`              | Return To Plane Probability                                            | [Scalar](#scalar)       | { `mapmri` }                                    | Probability \[0.0-1.0\]                                        |
| `tort`              | Tortuosity of extra-cellular space                                     | [Scalar](#scalar)       | { `dki` }                                       |                                                                |

While not explicitly included in the table above, *any* [Scalar](#scalar)
extrinsic parameter can theoretically be combined with a separate source
of orientation information from the diffusion model in order to produce a
[directionally-encoded colours (DEC)](#directionally-encoded-colours),
[spherical coordinates](#spherical-coordinates) or
[3-vectors](#3-vectors) image.

## Demonstrative examples

> -   A basic Diffusion Tensor fit:

Reviewer 3: This is great, but I think we should put it above

Author: Similar to some previous comments:

I've tried to make this more of a "specification" document and less of a "tutorial" document. Since many of the examples draw from content throughout the entirety of the rest of the document, it would be incomplete to present such prior to any of that content. If the slightly technical / general descriptions are confusing to a reader, they can simply "skip" to the end of the document to see the examples; I don't think it's our role here to hold the reader's hand closely as they try to familiarise themselves by reading through the document in a linear fashion. If there's a specific point that feels like it currently requires a demonstrative example to convey, we should probably look at it more closely; either improving the explanation, or with a minimal example that demonstrates just that one specific idea.
